paymentgateway/paypal/classes/ipn.php

```diff
@@ -0,0 +1,185 @@
+        global $CFG;
+
+        // Open a connection back to PayPal to validate the data.
+        $paypaladdr = empty($CFG->usepaypalsandbox) ? 'ipnpb.paypal.com' : 'ipnpb.sandbox.paypal.com';
+        $c = new \curl();
+        $options = array(
+            'returntransfer' => true,
+            'httpheader' => array('application/x-www-form-urlencoded', "Host: $paypaladdr"),
+            'timeout' => 30,
+            'CURLOPT_HTTP_VERSION' => CURL_HTTP_VERSION_1_1,
+        );
+        $location = "https://$paypaladdr/cgi-bin/webscr";
+        $result = $c->post($location, $this->req, $options);
+
+        if ($c->get_errno()) {
+            throw new \moodle_exception('errpaypalconnect', 'enrol_paypal', '', array('url' => $paypaladdr, 'result' => $result),
+                json_encode($data));
+        }
+
+        return $result;
+    }
+
+    /**
+     * Checks if anything is wrong with transaction data, and deals
+     * with errors by adding them to error_info.
+     *
+     * @param object $data
+     * @return bool $noerror
+     */
+    private function is_ipn_data_correct(&$data) {
+        global $DB;
+
+        $noerror = true;
+
+        $currency = get_config('paymentgateway_paypal', 'currency');
+        $cost = $DB->get_field('tool_paymentplugin_course', 'cost', array('courseid' => $data->courseid));
+
+        // Check that course price and currency matches.
+        $error_info = "";
+        if ($data->mc_currency != $currency) {
+            $error = false;
+            $error_info .= get_string('erroripncurrency', 'paymentgateway_paypal') . " ";
+        }
+        if ($data->mc_gross != $cost) {
+            $error = false;
+            $error_info .= get_string('erroripncost', 'paymentgateway_paypal') . " ";
+        }
+
+        // Check that courseid and userid are valid.
+        if (!$DB->record_exists('course', array('id' => $data->courseid))) {
+            $error = false;
+            $error_info .= get_string('erroripncourseid', 'paymentgateway_paypal') . " ";
+        }
+        if (!$DB->record_exists('user', array('id' => $data->userid))) {
+            $error = false;
+            $error_info .= get_string('erroripnuserid', 'paymentgateway_paypal') . " ";
+        }
+
+        if (!$noerror) {
+            // Leave record of unsuccessful purchase in database with error details.
+            $data->error_info = $error_info;
+        }
+
+        return $noerror;
+    }
+
+    /**
+     * This function sends the transaction data through to the gateway to be filtered and the transaction, actioned.
+     * 
+     * @param string $result 'INVALID' if there was an issue with the paypent.
+     * @param \stdclass $data The transaction data.
+     */
+    public function submit_data($result, $data) {
+        $paypalgateway = \tool_paymentplugin\plugininfo\paymentgateway::get_gateway_object('paypal');
+
+        if (strcmp ($result, "INVALID") == 0) {                 // INVALID PAYPAL IPN
```

Comment:
This invalid string seems a little too magic. I think move this to a class const